libs/qCC_db/src/ccRasterGrid.cpp

```diff
@@ -401,79 +292,197 @@ bool ccRasterGrid::fillWith(	ccGenericPointCloud* cloud,
+				if (projectionType == PROJ_AVERAGE_VALUE)
+				{
+					//we keep track of the point which is the closest to the cell center (in 2D)
+					CCVector2d C = computeCellCenter(cellPos.x, cellPos.y, X, Y);
+					double minimum_squareDistToP;
+					for (unsigned n=0; n < aCell.nbPoints ;n++)
+					{	
+						unsigned pointIndex = cellPointIndexedHeight[n].index;
+						const CCVector3* P = cloud->getPoint(pointIndex); 
+						CCVector2d P2D(P->u[X], P->u[Y]);
+						double squareDistToP = (C - P2D).norm2();
+						if ((squareDistToP < minimum_squareDistToP) || (n==0)){
+							minimum_squareDistToP=squareDistToP;
+							aCell.pointIndex = pointIndex;
 						}
 					}
 				}
-			}
-		}
-	}
 
-	//update the main grid (average height and std.dev. computation + current 'height' value)
-	{
-		for (unsigned j = 0; j < height; ++j)
-		{
-			Row& row = rows[j];
-			for (unsigned i = 0; i < width; ++i)
-			{
-				ccRasterCell& cell = row[i];
-				if (cell.nbPoints > 1)
+				// Handle if datasets has RGB-colors 
+				if (hasColors)
 				{
-					cell.avgHeight /= cell.nbPoints;
-					cell.stdDevHeight = sqrt(std::abs(cell.stdDevHeight / cell.nbPoints - cell.avgHeight*cell.avgHeight));
-					if (hasColors && projectionType == PROJ_AVERAGE_VALUE)
-					{
-						cell.color /= cell.nbPoints;
+					assert(cloud->hasColors());
+					if (projectionType == PROJ_AVERAGE_VALUE){
+						//Average Color
+						aCell.color = CCVector3d(0,0,0);
+						for (unsigned n=0; n < aCell.nbPoints ;n++){
+							unsigned pointIndex = cellPointIndexedHeight[n].index;
+							const ccColor::Rgb& col = cloud->getPointColor(pointIndex);
+							aCell.color += CCVector3d(col.r, col.g, col.b);
+						}
+						aCell.color /= aCell.nbPoints;
+					}
+					else{  
+						//Pick color from selected index
+						const ccColor::Rgb& col = cloud->getPointColor(aCell.pointIndex);
+						aCell.color = CCVector3d(col.r, col.g, col.b);
 					}
 				}
-				else
-				{
-					cell.stdDevHeight = 0;
-				}
-
-				if (cell.nbPoints != 0)
+				
+				//Handle scalar fields
+				if (interpolateSF)
 				{
-					//set the right 'height' value
-					switch (projectionType)
+					assert(pc);
+					//absolute position of the cell (e.g. in the 2D SF grid(s))
+					int pos = cellPos.y * static_cast<int>(width) + cellPos.x;
+					assert(pos < static_cast<int>(gridTotalSize));
+					
+					for (size_t k = 0; k < scalarFields.size(); ++k)
 					{
-					case PROJ_MINIMUM_VALUE:
-						cell.h = cell.minHeight;
-						break;
-					case PROJ_AVERAGE_VALUE:
-						cell.h = cell.avgHeight;
-						break;
-					case PROJ_MAXIMUM_VALUE:
-						cell.h = cell.maxHeight;
-						break;
-					default:
-						assert(false);
-						break;
+						assert(!scalarFields[k].empty());
+						CCCoreLib::ScalarField* sf = pc->getScalarField(static_cast<unsigned>(k));
+						assert(sf && pos < scalarFields[k].size());
+						// Set up vector of valid sf values for current cell 
+						cellPointSF.reserve(aCell.nbPoints);
```

Comment:
resize 😅 (does this code work in debug mode?)

Reply:
Comment on use of reserve vs resize in the main comment thread